### PR TITLE
xsJelmer gun

### PR DESCRIPTION
**Pull request by waseland**
Well okay ¯\\\_(ツ)_/¯ "Able to merge"

### CODE REVIEW

File: src/gamelogic/Game.cpp [hunk 1]
```diff
@@ -87,6 +87,7 @@ namespace GameLogic {
         }
 
         this->_player->set_level(this->_level);
+        this->_init_weapons();
```
Comment (jorgdebont) [inline src/gamelogic/Game.cpp]: Make a weapon base class as discussed

File: src/gamelogic/Game.cpp [hunk 2]
```diff
@@ -100,6 +101,58 @@ namespace GameLogic {
     void Game::_init_sound_effects() {
         this->SDL_facade.load_sound_effect("monsterkill", "res/sound_effects/monsterkill.wav");
         this->SDL_facade.load_sound_effect("headshot", "res/sound_effects/headshot.wav");
+        this->SDL_facade.load_sound_effect("pistolshot", "res/sound_effects/pistolshot.wav");
```
Comment (jorgdebont) [inline src/gamelogic/Game.cpp]: Sound effect should also be loaded in the new derived classes

File: src/gamelogic/Game.cpp [hunk 3]
```diff
@@ -100,6 +101,58 @@ namespace GameLogic {
...
+        this->SDL_facade.load_sound_effect("machinegunshot", "res/sound_effects/machinegunshot.wav");
+        this->SDL_facade.load_sound_effect("chaingunshot", "res/sound_effects/chaingunshot.wav");
+    }
+
+    void Game::_init_weapons()
```
Comment (jorgdebont) [inline src/gamelogic/Game.cpp]: Should all be done in derived classes

File: src/gamelogic/Player.cpp [hunk 4]
```diff
@@ -186,4 +193,44 @@ namespace GameLogic {
...
+            this->_weapons.push_back(weapon);
+        } else {
+            throw Exceptions::WeaponIsNullptrException();
+        }
+    }
+
+    void Player::_next_weapon()
+    {
+        if (!(this->get_weapon()->being_shot())) {
+            if (this->_current_weapon + 1 < (int) this->_weapons.size()) {
```
Comment (jorgdebont) [inline src/gamelogic/Player.cpp]: you can do this with module
index % weapons.size() = new index

File: src/gamelogic/Player.cpp [hunk 5]
```diff
@@ -186,4 +193,44 @@ namespace GameLogic {
...
+                this->_current_weapon++;
+            } else {
+                this->_current_weapon = 0;
+            }
+        }
+    }
+
+    void Player::_previous_weapon()
```
Comment (jorgdebont) [inline src/gamelogic/Player.cpp]: you can do this with module
index % weapons.size() = new index

File: src/gamelogic/Player.hpp [hunk 6]
```diff
@@ -64,6 +70,11 @@ namespace GameLogic {
         void _set_new_current_tile();
         void _try_trigger();
         void _do_action();
+        void _next_weapon();
+        void _previous_weapon();
+
+        vector<Weapon*> _weapons;
+        int _current_weapon = 0;
```
Comment (jorgdebont) [inline src/gamelogic/Player.hpp]: _current_weapon_index

File: src/gamelogic/Weapon.cpp [hunk 7]
```diff
@@ -0,0 +1,71 @@
+//
+// Created by waseland on 12/19/16.
+//
+
+#include "Weapon.hpp"
+
+namespace GameLogic {
+
+    Weapon::Weapon(SDLFacade& SDL_Facade, CoordinateInt draw_coordinates
```
Comment (jorgdebont) [inline src/gamelogic/Weapon.cpp]: if you hardcode the vars in the derived classes then perhaps you should remove them from init list

File: src/gamelogic/Weapon.cpp [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    bool Weapon::being_shot()
+    {
+        return this->_shoot;
+    }
+
+    int Weapon::get_amount_of_shots()
+    {
+        return this->_amount_of_shots;
```
Comment (jorgdebont) [inline src/gamelogic/Weapon.cpp]: amount of shots might not be the best variable name
perhaps bullets_per_shot or something

File: src/gamelogic/Weapon.hpp [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    public:
+        Weapon(SDLFacade& SDL_Facade, CoordinateInt draw_coordinates
+        , int animation_start_image, std::string sound_effect, int image_interval
+        , std::vector<std::string> images, int amount_of_shots, int damage_per_shot);
+        virtual ~Weapon();
+
+        void draw() override;
+
+        void shoot();
+        bool being_shot();
```
Comment (jorgdebont) [inline src/gamelogic/Weapon.hpp]: Do we ever need this?

File: src/gamelogic/Player.cpp [hunk 10]
```diff
@@ -186,4 +193,44 @@ namespace GameLogic {
...
+        } else {
+            throw Exceptions::WeaponIsNullptrException();
+        }
+    }
+
+    void Player::give_weapon(Weapon* weapon) {
+        if (weapon != nullptr) {
+            this->_weapons.push_back(weapon);
+        } else {
+            throw Exceptions::WeaponIsNullptrException();
```
Comment (jorgdebont) [inline src/gamelogic/Player.cpp]: Exception is never caught

File: src/gamelogic/Player.cpp [hunk 11]
```diff
@@ -186,4 +193,44 @@ namespace GameLogic {
             }
         }
     }
+
+    Weapon *Player::get_weapon() {
+        if(this->_weapons[this->_current_weapon] != nullptr){
+            return this->_weapons[this->_current_weapon];
+        } else {
+            throw Exceptions::WeaponIsNullptrException();
```
Comment (jorgdebont) [inline src/gamelogic/Player.cpp]: Exception is never caught

File: src/gamelogic/Weapon.cpp [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+                this->_ticks_on_shoot = this->_previous_ticks;
+            }
+            if (this->_current_image >= (int)this->_images.size()) {
+                this->_shoot = false;
+                this->_current_image = this->_animation_start_image;
+            }
+        } else {
+            this->_SDL_facade.draw_image(this->_images[0], this->_draw_coordinates);
+        }
+        _previous_ticks = this->_SDL_facade.get_ticks();
```
Comment (jorgdebont) [inline src/gamelogic/Weapon.cpp]: this->
prob missed more of them

File: src/gamelogic/state/RunState.cpp [hunk 13]
```diff
@@ -42,6 +42,7 @@ namespace State {
 
         this->_context.raycasting_engine.draw();
         this->_collection.draw();
+        this->_context.get_level()->get_player().get_weapon()->draw();
```
Comment (jorgdebont) [inline src/gamelogic/state/RunState.cpp]: should be added to the collection instead

File: src/gamelogic/Game.cpp (hunk 1 above)
Comment (waseland) [inline src/gamelogic/Game.cpp]: Been there done that 🐶

File: src/gamelogic/state/RunState.cpp (hunk 13 above)
Comment (waseland) [inline src/gamelogic/state/RunState.cpp]: Doesn't work, weapon needs to be changed in runtime and can't be put as a drawable in the collection as it is now

File: src/gamelogic/weapons/ChainGun.cpp [hunk 14]
```diff
@@ -0,0 +1,37 @@
+//
+// Created by waseland on 12/20/16.
+//
+
+#include "ChainGun.hpp"
+
+namespace GameLogic {
+
+    ChainGun::ChainGun(SDLFacade& SDL_Facade)
+            : Weapon(SDL_Facade)
```
Comment (jorgdebont) [inline src/gamelogic/weapons/ChainGun.cpp]: shift + tab

File: src/gamelogic/Weapon.cpp (hunk 12 above)
Comment (jorgdebont) [inline src/gamelogic/Weapon.cpp]: @waseland you missed this

File: src/gamelogic/Weapon.cpp (hunk 8 above)
Comment (jorgdebont) [inline src/gamelogic/Weapon.cpp]: @waseland you missed this

File: src/gamelogic/Weapon.hpp (hunk 9 above)
Comment (jorgdebont) [inline src/gamelogic/Weapon.hpp]: @waseland you missed this

File: src/gamelogic/weapons/Weapon.cpp [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    {
+    }
+
+    Weapon::~Weapon()
+    {
+    }
+
+    void Weapon::draw()
+    {
+        if (_shoot == true) {
```
Comment (jorgdebont) [inline src/gamelogic/weapons/Weapon.cpp]: this

File: src/gamelogic/weapons/MachineGun.cpp [hunk 16]
```diff
@@ -0,0 +1,36 @@
+//
+// Created by waseland on 12/20/16.
+//
+
+#include "MachineGun.hpp"
+
+namespace GameLogic {
+
+    MachineGun::MachineGun(SDLFacade& SDL_Facade)
+            : Weapon(SDL_Facade)
```
Comment (jorgdebont) [inline src/gamelogic/weapons/MachineGun.cpp]: shift tab

File: src/gamelogic/weapons/Weapon.cpp (hunk 15 above)
Comment (jorgdebont) [inline src/gamelogic/weapons/Weapon.cpp]: this

File: src/gamelogic/Player.cpp [hunk 17]
```diff
@@ -195,4 +202,39 @@ namespace GameLogic {
...
+        }
+    }
+
+    void Player::_next_weapon()
+    {
+        if (!(this->get_weapon()->being_shot())) {
+            this->_current_weapon_index++;
+            this->_current_weapon_index = this->_current_weapon_index % (int) this->_weapons.size();
```
Comment (jorgdebont) [inline src/gamelogic/Player.cpp]: this->_current_weapon_index = (this->_current_weapon_index < 0) ? this->_current_weapon_index + this->_weapons.size() : this->_current_weapon_index;

Comment (jorgdebont) [inline src/gamelogic/Player.cpp]: Apparently % isn't actually modulo, but is remainder

Comment (jorgdebont) [inline src/gamelogic/Player.cpp]: so you have to add this below the line

File: src/gamelogic/Player.cpp [hunk 18]
```diff
@@ -195,4 +202,39 @@ namespace GameLogic {
...
+        }
+    }
+
+    void Player::_previous_weapon()
+    {
+        if (!(this->get_weapon()->being_shot())) {
+            this->_current_weapon_index--;
+            this->_current_weapon_index = this->_current_weapon_index % (int) this->_weapons.size();
```
Comment (jorgdebont) [inline src/gamelogic/Player.cpp]: this->_current_weapon_index = (this->_current_weapon_index < 0) ? this->_current_weapon_index + this->_weapons.size() : this->_current_weapon_index;

same here ofcourse

File: src/gamelogic/Player.cpp (hunk 17 above)
Comment (waseland) [inline src/gamelogic/Player.cpp]: It's magical that it worked 🔎

**Comment by waseland** [comment]
❤️

**Pull request merged**